### ISSUE TITLE
#### Issue Some CBZs do not open correctly.

### ISSUE THREAD

**Issue by DarkRookie00** [koreader/koreader#11304]
* KOReader version: 2023.10
* Device: Android Phone, Android Tablet, Linux AppImage

#### Issue
Some CBZs do not open correctly. All the items open stretched out. Extracting from the CBZ to a folder, and then opening the JPGs have the same issue. Change the extension from CBZ to ZIP doesn't change the item. Opening the JPG in a picture editing has the image load normally. The issue also occurs with  the Librera app on Android. Doesn't with the Book Reader app. Doesn't occur on my computer using Okular and GwenView.

#### Steps to reproduce
Open a CBZ file to see if issue occurs.
[crash.log](https://github.com/koreader/koreader/files/13797115/crash.log)

**Comment by NiLuJe** [comment]
Probably going to need a testcase to repro. Since we're talking JPEG, a random guess would be funky exif or something.

**Comment by hugleo** [comment]
I created a test case.

![test](https://github.com/koreader/koreader/assets/4573505/3ca0af23-bf56-4471-bad5-58b025217934)


The KoReader ball gets stretched in KoReader.
![koreader](https://github.com/koreader/koreader/assets/4573505/eb0babfe-e80b-488d-bb11-263a06c2dbc4)


The KoReader ball doesn't get stretched in Okular.
![okular](https://github.com/koreader/koreader/assets/4573505/a3c97a0e-6023-4902-a310-2463f15a34ad)

**Comment by hugleo** [comment]
Same problem when testing directly on mupdf 1.23.7.

If you open it in KoReader using the Picture Document or Image viewer, the image looks normal.

**Comment by hugleo** [comment]
The issue is really with the JPEG format. It looks normal in MuPDF/KoReader when I convert the image to PNG.

**Comment by Frenzie** [comment]
I think you left out a very major component (the actual image). ;-)

**Comment by hugleo** [comment]
It's based on similar characteristics with other pixels. I checked that the stretch thing works in a similar manner. 
But you are right, it's not exactly the OP image. Other test cases here are welcome!

![demensions](https://github.com/koreader/koreader/assets/4573505/907d30da-e44b-43a8-8e94-39226bbd85eb)

**Comment by DarkRookie00** [comment]
What [hugleo](https://github.com/hugleo) posted is exactly what I am seeing.

**Comment by Frenzie** [comment]
> It's based on similar characteristics with other pixels. I checked that the stretch thing works in a similar manner. But you are right, it's not exactly the OP image. Other test cases here are welcome!

Ok, but what do you have against sharing yours? :p

**Comment by hugleo** [comment]
> > It's based on similar characteristics with other pixels. I checked that the stretch thing works in a similar manner. But you are right, it's not exactly the OP image. Other test cases here are welcome!
> 
> Ok, but what do you have against sharing yours? :p

Haven't I already shared the message at https://github.com/koreader/koreader/issues/11304#issuecomment-1872501932 ? 
I think I understand what you mean. 
I see issues on GitHub when trying to download the image before, but when I tried with another browser, it worked.

**Comment by Frenzie** [comment]
No, I'm typing on my phone and I just didn't notice the first image was the actual file and not a screenshot. Though I wouldn't expect it to be embedded regardless.

**Comment by NiLuJe** [comment]
My best guess is that's caused by the the 4:2:0 subsampling (quartered chroma):

`jpeg:sampling-factor: 2x2,1x1,1x1`  vs., ` jpeg:sampling-factor: 1x1,1x1,1x1` w/ full chroma (4:4:4).

**Comment by hugleo** [comment]
Seems like this image has a defect in the header, returning the same width/height for X_density/Y_density (1100/1650).

```bash
$ python image_info.py defect.jpg 
Image Information:
image_width: 1100
image_height: 1650
X_density: 1100
Y_density: 1650

$ python image_info.py normal.jpg 
Image Information:
image_width: 1100
image_height: 1650
X_density: 96
Y_density: 96
```

In MuPDF, the values are gathered in this function:

[MuPDF Load JPEG Function](https://github.com/ArtifexSoftware/mupdf/blob/master/source/fitz/load-jpeg.c#L409)

I added the following code at the end of this function just for testing purposes, and the image was loaded without stretching:

```c
if (*xresp == *xp && *yresp == *yp) {
    *xresp = 96;
    *yresp = 96;
}
```

**Comment by DarkRookie00** [comment]
I found another couple CBZ files that straight wont work in this. Will work in normal image programs. Same thing with the images all being JPGs

**Comment by hugleo** [comment]
Another two cents here. I think exiftool can be used to change the DPI header of the image.

```
$ exiftool -XResolution -YResolution -ResolutionUnit image.jpg 
X Resolution                    : 1100
Y Resolution                    : 1650
Resolution Unit                 : inches

$ exiftool -XResolution=96 -YResolution=96 -ResolutionUnit=inches image.jpg 
    1 image files updated
    
$ exiftool -XResolution -YResolution -ResolutionUnit image.jpg 
X Resolution                    : 96
Y Resolution                    : 96
Resolution Unit                 : inches
```

**Comment by Frenzie** [comment]
The behavior seems to be exactly as expected for both JFIF and EXIF. I can only assume some other software would effectively be broken on pictures with a different x and y resolution.
https://en.wikipedia.org/wiki/JPEG_File_Interchange_Format#JFIF_APP0_marker_segment

**Comment by DynamoRocket** [comment]
Okular and Gwenview on PC don't have this issue. Those open the files without issue.

**Comment by hugleo** [comment]
As mentioned in https://github.com/koreader/koreader/issues/11304#issuecomment-1873016826, mupdf honors what's in the header, which should be the correct behavior. Other software may not honor the density or may use workarounds, as the one in the same comment.
The best option is to fix the images, for example, by using a script with exiftool on all images.

**Comment by Frenzie** [comment]
> Those open the files without issue.

Which is the broken behavior since they should be scaled as they ask. :-) However, I'm not quite sure what specifying a different x and y density would be good for. I can only assume it must've been useful in the (early) '90s as a way to squeeze out a few extra bytes.

**Comment by DynamoRocket** [comment]
Then why does it stretch to the full screen and not honor the aspect ratio?

**Comment by hugleo** [comment]
I guess you "lose" quality when applying the sampling algorithm with wrong density.

**Comment by DynamoRocket** [comment]
Off to find a different app then. I am not capable of doing that across 1000 files that I know of.

**Comment by Frenzie** [comment]
> What in this would make it stretch?

Because it's saying in the horizontal direction it has a density of 1100 pixels per inch and in the vertical direction it has a density of 1650 pixels per inch.

Combined with the resolution of 1100 × 1650 that means it's a square picture of 1 inch by 1 inch.

Typical values include 72 × 72, 96 × 96, 150 × 150, 300 × 300, and a bit less typically 600 × 600 and 1200 × 1200.

If we take a more typical 96 × 96 as an example, that means physically it would be (1100 / 96) × (1650 / 96) = 11.5 inch × 17.1 inch, which sounds in the right ballpark for a comic book, but it might indicate it's actually closer to 150 × 150 PPI.
